text/0000-from-lossy.md

```diff
@@ -0,0 +1,416 @@
+range, this conversion trait is still applicable.)
+
+The implementations should fail on NaN, Inf, and values whose integer
+approximation is not representable. The integer approximation should be the
+value rounded towards zero. E.g.:
+
+- 1.6f32 → u32: 1
+- -1f32 → u32: error
+- -0.2f32 → u32: 0
+- 100_000f32 → u16: error
+
+## Other conversions
+
+### `WrappingFrom` trait
+
+There are several types of integer conversion which do not preserve the
+semantic value:
+
+-   sign transmutation (i.e. reinterpretation as another integer type of the
+    same size but with different signed-ness), e.g. `128u8 → -128i8`
+-   unsigned truncation (simply dropping excess high bits), e.g. `260u16 → 4u8`
+-   signed truncation (dropping excess high bits and sign-transmuting the
+    remainder), e.g. `-260i16 → -4i8`
+
+Note that there is less incentive against usage of `as` in these cases since the
+conversions do not preserve "value", although alternatives still have some use
+(e.g. to clarify that a conversion is a truncation).
+
+We can add `std::convert::WrappingFrom`:
+
+```rust
+/// A trait for wrapping conversions; these may truncate (drop high bits) and
+/// sign-convert (re-interpret meaning of high bits).
+/// 
+/// Note that `TryFrom` and `WrappingFrom` may both be implemented for the same
+/// source and target types; in this case they should have equivalent result
+/// where `TryFrom` succeeds.
+pub trait WrappingFrom {
+    fn wrapping_from(x: T) -> Self;
+}
+```
+
+Add implementations for all integer conversions which are *not implemented* by
+`From`. These conversions should truncate or zero-extend to the size of the
+result type, then reinterpret (transmute) the bits to the result type.
+
+Potentially this could also re-implement all `From` conversions on integer
+types. This could be useful for generic code. Unfortunately since Rust has no
+`Integer` trait there is no simple way to re-implement `From` *only* where the
+source and target types are integer types.
+
+# Related problems
+
+These problems are discussed in the search for a complete solution; however it
+is not currently proposed to solve them within this RFC.
+
+## Platform-dependent types
+
+### isize / usize
+
+The `isize` and `usize` types have undefined size, though there appears to be
+an assumption that they are at least 16 bits (existing `From` implementations)
+and that they could be larger than 64 bits.
+[Discussion thread on internals](https://internals.rust-lang.org/t/numeric-into-should-not-require-everyone-to-support-16-bit-and-128-bit-usize/3609/7).
+[Discussion about `usize` → `u64` on users forum](https://users.rust-lang.org/t/cant-convert-usize-to-u64/6243).
+
+Checked integer conversions using `TryFrom` [are being reintroduced](https://github.com/rust-lang/rust/issues/49415).
+
+It is possible that unchecked conversions could be added, perhaps using
+`WrappingFrom` (or some other trait allowing both truncation and
+zero-extension).
```

Comment:
I would absolutely expect `usize` and `isize` to be in the cartesian product of numeric types for `WrappingFrom`, since something like `u32::wrapping_from(-1_isize)` is perfectly well-defined.  Yes, `usize::wrapping_from(-1)` will have platform-dependent behaviour, but it's well-defined so no worse than `usize::MAX`.